// eslint-plugin-flowtype-errors/dist/get-program.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (source, node) {

  // Ignore if body is empty.
  if (node.body.length === 0) {
    return;
  }

  var body0 = node.body[0];
  var comments0 = node.comments[0];
  var start = void 0;
  var range = void 0;

  // With babel-eslint, program.loc.start.column is always 0,
  // workaround it by using the first node of the body to get the offset.

  if (comments0) {
    start = node.range[0] < comments0.range[0] ? {
      line: body0.loc.start.line,
      column: body0.loc.start.column
    } : {
      line: comments0.loc.start.line,
      column: comments0.loc.start.column
    };
    range = [Math.min(node.range[0], comments0.range[0]), Math.max(node.range[1], node.comments[node.comments.length - 1].range[1])];
  } else {
    start = {
      line: body0.loc.start.line,
      column: body0.loc.start.column
    };
    range = node.range;
  }

  return {
    text: source.text.slice(range[0], range[1]),
    loc: {
      start,
      end: start
    },
    offset: {
      line: start.line - 1,
      column: start.column
    }
  };
};

module.exports = exports["default"];
